### PR TITLE
feat: impl alter table in distributed mode

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
- Implement alter table in distributed mode
- Rename `common/insert` crate to `common/gprc-expr`
- Move some gRPC struct to request convesion functions to `common/gprc-expr`

## Checklist

- [X]  I have written the necessary rustdoc comments.
- [X]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
NA

### CODE REVIEW

**Comment by evenyag** [comment]
Could you please resolve the conflict first?

File: src/frontend/src/instance/distributed.rs [hunk 1]
```diff
@@ -166,6 +169,34 @@ impl DistInstance {
...
+            .context(CatalogSnafu)?
+            .context(TableNotFoundSnafu {
+                table_name: format!("{}.{}.{}", catalog_name, schema_name, table_name),
+            })?;
+
+        let dist_table = table
+            .as_any()
+            .downcast_ref::<DistTable>()
+            .expect("Table impl must be DistTable in distributed mode");
+        dist_table.alter_by_expr(expr).await.unwrap();
```
Comment (MichaelScofield) [inline src/frontend/src/instance/distributed.rs:196]: Is unwrap safe here?

File: src/frontend/src/table.rs [hunk 2]
```diff
@@ -348,6 +352,26 @@ impl DistTable {
         };
         Ok(partition_rule)
     }
+
+    /// Define a `alter_by_expr` instead of impl [`Table::alter`] to avoid redundant conversion between  
+    /// [`table::requests::AlterTableRequest`] and [`AlterExpr`].
+    pub(crate) async fn alter_by_expr(&self, expr: AlterExpr) -> Result<()> {
```
Comment (MichaelScofield) [inline src/frontend/src/table.rs:358]: I think we should implement DistTable::alter and use it. AlterTable -> AlterTableRequest -> AlterExpr for DistTable is fine, doesn't have to "must use" the `alter_expr` in Frontend instance [here](https://github.com/GreptimeTeam/greptimedb/pull/572/files#diff-7b73fa585fc3ed84f34902a8126a169043dcf771cb9a2961d3795bab34db221eR272)

File: src/frontend/src/table.rs [hunk 3]
```diff
@@ -348,6 +352,26 @@ impl DistTable {
...
+        let table_routes = self.table_routes.get_route(&self.table_name).await.unwrap();
```
Comment (MichaelScofield) [inline src/frontend/src/table.rs:359]: Is unwrap safe here?

File: src/frontend/src/table.rs [hunk 4]
```diff
@@ -348,6 +352,26 @@ impl DistTable {
...
+        for datanode in table_routes.find_leaders() {
```
Comment (MichaelScofield) [inline src/frontend/src/table.rs:360]: If "leaders" are empty, I think we'd better return an error. After all, the "alter" is not actually applied.

File: src/frontend/src/table.rs (hunk 2 above)
Comment (v0y4g3r) [reply src/frontend/src/table.rs:358]: When it comes to alter table on insertion, it would be `AlterExpr`->`AlterTableRequest`->`AlterExpr`

**Review by MichaelScofield — APPROVED** [review]

**Review by waynexia — APPROVED** [review]
Looks good to me 👍

File: src/frontend/src/instance/distributed.rs [hunk 5]
```diff
@@ -166,6 +169,34 @@ impl DistInstance {
...
+        let table_name = expr.table_name.as_str();
+        let table = self
+            .catalog_manager
+            .catalog(catalog_name)
+            .context(CatalogSnafu)?
+            .context(CatalogNotFoundSnafu { catalog_name })?
+            .schema(schema_name)
+            .context(CatalogSnafu)?
+            .context(SchemaNotFoundSnafu {
+                schema_info: format!("{}.{}", catalog_name, schema_name),
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/frontend/src/instance/distributed.rs:184]: I'm ok with it as long as it won't cause enum variant to become too large.

File: src/common/grpc-expr/src/alter.rs [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+use crate::error::{
+    ColumnNotFoundSnafu, CreateSchemaSnafu, InvalidColumnDefSnafu, MissingFieldSnafu,
+    MissingTimestampColumnSnafu, Result,
+};
+
+/// Convert an [`AlterExpr`] to an optional [`AlterTableRequest`]
+pub fn alter_expr_to_request(expr: AlterExpr) -> Result<Option<AlterTableRequest>> {
+    match expr.kind {
+        Some(Kind::AddColumns(add_columns)) => {
+            let mut add_column_requests = vec![];
```
Comment (killme2008) [inline src/common/grpc-expr/src/alter.rs:34]: Vec::with_capacity(add_columns.add_columns.len());

File: src/frontend/src/table.rs [hunk 7]
```diff
@@ -348,6 +352,36 @@ impl DistTable {
...
+                    expr.catalog_name, expr.schema_name, expr.table_name
+                )
+            }
+        );
+        for datanode in leaders {
+            let admin = Admin::new(
+                DEFAULT_CATALOG_NAME,
+                self.datanode_clients.get_client(&datanode).await,
+            );
+            info!("Sent alter table {:?} to {:?}", expr, admin);
```
Comment (killme2008) [inline src/frontend/src/table.rs:375]: I think `debug` is enough.

File: src/frontend/src/table.rs [hunk 8]
```diff
@@ -348,6 +352,36 @@ impl DistTable {
...
+            let result = admin
+                .alter(expr.clone())
+                .await
+                .context(RequestDatanodeSnafu)?;
+            info!("Alter table result: {:?}", result);
```
Comment (killme2008) [inline src/frontend/src/table.rs:380]: Not necessary to use `info`. The log would be annoying.

**Review by killme2008 — APPROVED** [review]

File: src/common/grpc-expr/src/alter.rs (hunk 6 above)
Comment (killme2008) [reply src/common/grpc-expr/src/alter.rs:34]: But I would prefer to collect from the iterator.

**Pull request merged**